workflow/controller/retry_tweak.go

```diff
@@ -11,13 +11,30 @@ import (
 // RetryTweak is a 2nd order function interface for tweaking the retry
 type RetryTweak = func(retryStrategy wfv1.RetryStrategy, nodes wfv1.Nodes, pod *apiv1.Pod)
 
-// FindRetryNode locates the closes retry node ancestor to nodeID
-func FindRetryNode(nodes wfv1.Nodes, nodeID string) *wfv1.NodeStatus {
+// FindRetryNode locates the closes retry node ancestor to nodeID.
+// If boundaryID is set on the node we get the template name of the boundary node and try to find a node of type retry with
+// the same template name
+// If boundaryID is no set we get the template of the node and try to find a node of type retry with the name name as the template
+func (woc *wfOperationCtx) FindRetryNode(nodes wfv1.Nodes, nodeID string) *wfv1.NodeStatus {
 	boundaryID := nodes[nodeID].BoundaryID
-	boundaryNode := nodes[boundaryID]
-	templateName := boundaryNode.TemplateName
+
+	if boundaryID != "" {
+		boundaryNode := nodes[boundaryID]
+		templateName := boundaryNode.TemplateName
```

Comment:
I suppose this will panic if nodes[boundaryID] isn't found, so you could add error handling for that (on the other hand, I know that would indicate a very serious problem in that case, and maybe we should panic) - will let others decide

Reply:
I think it would be better to have the error handled gracefully rather than kicking a panic. We could do:
```
node, ok := nodes[nodeID]
if !ok {
  // return node not found error
}

boundaryID := node.BoundaryID
...
```